theano/tensor/sort.py

```diff
@@ -206,3 +219,311 @@ def argsort(a, axis=-1, kind='quicksort', order=None):
+                dtype=self.idx_dtype, broadcastable=bcast)())
+        return theano.Apply(self, [inp, kth], outs)
+
+    def perform(self, node, inputs, output_storage):
+        x, k = inputs
+        axis = self.axis
+        if not self.return_indices:
+            pzv = output_storage[0]
+            pzv[0] = _topk_py_impl(self, x, k, axis, None)
+        elif self.return_values:
+            pzv = output_storage[0]
+            pzi = output_storage[1]
+            pzv[0], pzi[0] = _topk_py_impl(
+                self, x, k, axis, node.outputs[1].dtype)
+        else:
+            pzi = output_storage[0]
+            pzi[0] = _topk_py_impl(self, x, k, axis, node.outputs[0].dtype)
+
+    def infer_shape(self, node, inp_shapes):
+        _check_tensor_is_scalar(node.inputs[1])
+        shp = list(inp_shapes[0])
+        shp[self.axis] = np.abs(node.inputs[1])
+        shp = tuple(shp)
+        return [shp for i in [self.return_values, self.return_indices] if i]
+
+    def L_op(self, inputs, outputs, out_grads):
+        x, k = inputs
+        k_grad = grad_undefined(self, 1, k, 'topk: k is not differentiable')
+        if not (self.return_indices, self.return_values):
+            x_grad = grad_undefined(
+                self, 0, x, 'topk: cannot get gradient'
+                ' without both indices and values')
+        elif x.ndim == 1:
+            z_grad = out_grads[0]
+            indices = outputs[-1]
+            x_grad = x.zeros_like(dtype=z_grad.dtype)
+            x_grad = theano.tensor.advanced_set_subtensor1(
+                x_grad, z_grad, indices)
+        else:
+            x_shp = theano.tensor.shape(x)
+            z_grad = out_grads[0]
+            ndim = x.ndim
+            axis = self.axis % ndim
+            grad_indices = [
+                arange(x_shp[i]).dimshuffle([0] + ['x'] * (ndim - i - 1))
+                if i != axis else outputs[-1] for i in range(ndim)]
+            x_grad = x.zeros_like(dtype=z_grad.dtype)
+            x_grad = theano.tensor.advanced_set_subtensor(
+                x_grad, z_grad, *grad_indices)
+        return [x_grad, k_grad]
+
+
+def topk(x, kth, axis=-1, idx_dtype='int64'):
+    """
+    Returns the k-largest elements along an axis.
+
+    Parameters
+    ----------
+
+    x: tensor instance
+
+    kth: integer constant/variable
+        Must not be 0. If negative, gives k-smallest elements instead.
+
+    axis: integer or ``None``
+        Upon which axis shall the operation be performed on.
+        If ``None``, works on flattened array.
+
+    idx_dtype: string
+        Specify output dtype used in indices, defaults to ``int64``, must be integer type.
+        This option is here because indices are needed for gradient.
+
+    Returns
+    -------
+    Tensor variable with same dtype as `x`.
+
+    Notes
+    -----
+    - The returned values are not sorted.
+
+    """
+    if axis is None:
+        x = theano.tensor.flatten(x)
+        axis = -1
```

Comment:
I would set the axis to 0 in that case. It would make the modulo not needed and this is more clear when the op get printed to the user I think. Idem for the others.